### ISSUE TITLE
"Saving 'file.tf': Running 'Hashicorp Terraform' Formatter" - takes forever to run

### ISSUE THREAD

**Issue by scretu** [hashicorp/vscode-terraform#1356]
### Extension

```
HashiCorp Terraform: v2.25.4
```

### VS Code

```
Version: 1.76.1
Release: 23069
Commit: 9561d503ffa83a1498d030540c4579e75c686489
Date: 2023-03-10T18:21:47.706Z
Electron: 19.1.9
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Linux x64 5.15.0-67-generic
Sandboxed: No
```

### Operating System

```
Ubuntu 22.04.2 LTS
```

### Terraform Version

```
$ terraform version
Terraform v1.2.9
on linux_amd64
```

### Steps To Reproduce

Basically every time I click Save (or press Ctrl+S) on a .tf file, I see the notification in the bottom right corner saying ""Saving 'file.tf': Running 'Hashicorp Terraform' Formatter" and it takes from a couple of seconds to... so many seconds that I have to click "Cancel"

### Expected Behavior

The file should be formatted and saved in an instant

### Actual Behavior

It takes from a couple of seconds to... so many seconds that I have to click "Cancel"

### Additional context

I saw many instances of this bug being reported... all of them are closed, but the issue is still present, it seems

**Comment by avgalani** [comment]
I expect it's the same issue as https://github.com/hashicorp/vscode-terraform/issues/1265; recommended current workaround is to downgrade to v2.23.0

**Comment by scretu** [comment]
@avgalani , unfortunaly, on v2.23.0, it is still slow :(

**Comment by reed-hanger** [comment]
Just adding on, I've had this issue in Windows pretty consistently for ~3 years now, throughout different versions of VSCode, Terraform, and the Terraform extension. It's pretty infuriating, It's always terraform-ls getting held up, and it tends to act the worst when I use the search find/replace (Ctrl + Shift + H), as it saves multiple files at the same time. It will literally run for hours if I don't stop it. *Sometimes* restarting VSCode helps, but usually I just disable Save on Format and just run `terraform fmt` in my folder, which processes all the files in less than 150ms. I've given up on it ever being fixed at this point.

**Comment by jseiser** [comment]
Having this issue on the normal, and beta release of this module.  

terraform-lsp behaves fine on neovim, so I assume its something vscode related.

We do use large monorepo's, and ive tried to make sure i only open vscode in as far down of  dir as possible, but hasnt improved anything.

**Comment by dbanck** [comment]
Thank you for taking the time to report this issue.

We released the new extension [2.26.0](https://github.com/hashicorp/vscode-terraform/releases/tag/v2.26.0) version (and [0.31.1](https://github.com/hashicorp/terraform-ls/releases/tag/v0.31.1) language server version) with [a potential fix](https://github.com/hashicorp/terraform-ls/pull/1258). So please try it and let us know if it solves this issue.

**Comment by 27Bslash6** [comment]
Same behaviour, latest version.  One reasonably large Terraform module with many (100) resources and submodules.

Sometimes (erratically) takes up to 20 minutes to save a file.

**Comment by radeksimko** [comment]
We plan to make some changes as part of https://github.com/hashicorp/terraform-ls/issues/1056 to help us understand situations like this (performance). While CPU/memory profiles can already be collected, these would not be as useful without context (such as some understanding of the workspace - no of files, folders, sizes of files etc.).

In the meantime, the only suggestion I can make is to avoid opening large/deep workspaces with lots of folders, esp. monorepos - i.e. open the one or few individual folders that you actually need to edit at any given time. This may not be most convenient, but should help avoid most performance issues for now.

Thank you for your patience and understanding.

**Comment by adrianisk** [comment]
I had to disable format on save because it consistently took 5+ seconds to save a file, even in a relatively small repo. While I was trying to figure out what was going on, I noticed that the extension enqueues some work on _every keystroke within a .tf file_. This is without even saving the file, you can watch the output as you type...

```
2023/06/07 16:30:42 opts.go:203: Received request batch of size 1 (qlen=60)
2023/06/07 16:30:43 opts.go:203: Received request batch of size 1 (qlen=61)
2023/06/07 16:30:43 opts.go:203: Received request batch of size 1 (qlen=62)
2023/06/07 16:30:43 opts.go:203: Received request batch of size 1 (qlen=63)
2023/06/07 16:30:44 opts.go:203: Received request batch of size 1 (qlen=64)
2023/06/07 16:30:45 opts.go:203: Received request batch of size 1 (qlen=65)
2023/06/07 16:30:45 opts.go:203: Received request batch of size 1 (qlen=66)
2023/06/07 16:30:45 opts.go:203: Received request batch of size 1 (qlen=67)
2023/06/07 16:30:45 opts.go:203: Received request batch of size 1 (qlen=68)
2023/06/07 16:30:46 opts.go:203: Received request batch of size 1 (qlen=69)
2023/06/07 16:30:46 opts.go:203: Received request batch of size 1 (qlen=70)
2023/06/07 16:30:46 opts.go:203: Received request batch of size 1 (qlen=71)
2023/06/07 16:30:46 opts.go:203: Received request batch of size 1 (qlen=72)
2023/06/07 16:30:46 opts.go:203: Received request batch of size 1 (qlen=73)
2023/06/07 16:30:47 opts.go:203: Received request batch of size 1 (qlen=74)
2023/06/07 16:30:47 opts.go:203: Received request batch of size 1 (qlen=75)
2023/06/07 16:30:47 opts.go:203: Received request batch of size 1 (qlen=76)
2023/06/07 16:30:47 opts.go:203: Received request batch of size 1 (qlen=77)
2023/06/07 16:30:47 opts.go:203: Received request batch of size 1 (qlen=78)
2023/06/07 16:30:47 opts.go:203: Received request batch of size 1 (qlen=79)
```

**Comment by staranto** [comment]
I know "me too" doesn't help, but... me too.  So frustrating.

> uname
Linux pop-os 6.2.6-76060206-generic #202303130630~1689015125~22.04~ab2190e SMP PREEMPT_DYNAMIC Mon J x86_64 x86_64 x86_64 GNU/Linux
> code --version
1.81.0
6445d93c81ebe42c4cbd7a60712e0b17d9463e97
x64
> terraform-ls --version
0.31.1

Also happens in an Intel Mac w/same versions.

I've downgraded, upgraded, uninstalled, reinstalled so many times over the last year that I've completely lost track.  I have a small root project that is sourcing a private registry module.  That module has 5 sub-modules, each of which have 1-3 .tf's and are creating no more than ~5 resources on each module {} call.  Collectively the root project references one of the child modules ~20x (iow... the root project has ~20 module { source = "<<url>>//submod" }'s).  I don't know if that's considered deeply nested, but it's deeper than some.  

Perhaps LOC is a useful metric to get a sense of size and complexity...

The root project has 11 .tf's and 512 LOC.
The referenced module (with the 5 sub-modules) has 16 .tf's and 505 LOC.

Doesn't seem too overwhelming.  I will say there is some pretty hairy complex statements using functions, conditional operations, for loops, etc to build out locals.  All of those are in the module.  Again, a vague bullet point, no idea if it's relevant.

My observation is that Intellisense seems to perform reasonably in that it pops quick enough that I don't particularly notice.  But file save with format on save is horrendously slow.  A simple, 10 line .tf might take >10 minutes.  I would say saving a file is never slower than ~10 seconds and often nears 20 minutes if I have enough patience to let it ride.

**Comment by jpogran** [comment]
We've released `2.27.2` which contains two fixes (https://github.com/hashicorp/terraform-ls/pull/1369, https://github.com/hashicorp/terraform-ls/pull/1372) in our continuing efforts to address this issue.

With the fix, we should be back to terraform-ls `v0.29.2` levels of CPU usage while maintaining the memory improvements of `v0.29.3`.

If you have the time, please give it a try and let us know how it works for you. Please open a new issue rather than replying here, as this issue has gotten so many replies with different experiences it is hard to process. 

I am going to close this as we're using https://github.com/hashicorp/vscode-terraform/issues/1557 as a central tracking issue, so subscribe to that to see continued updates.
